src/compiler/checker.ts

```diff
@@ -8172,9 +8187,26 @@ namespace ts {
         }
 
         function filterType(type: Type, f: (t: Type) => boolean): Type {
-            return type.flags & TypeFlags.Union ?
-                getUnionType(filter((<UnionType>type).types, f)) :
-                f(type) ? type : neverType;
+            if (!(type.flags & TypeFlags.Union)) {

```

Comment:
The rewrite here makes sense (I'm assuming this is a perf thing?) but deserves a comment as to why it isn't using the existing filter code so that a maintainer doesn't just undo this change later.